### PR TITLE
Add Plugin search-backend-module-api-docs

### PR DESCRIPTION
**Pull request by jolavb**
## PR For Adding plugin-search-backend-module-api-docs

This is an extension to the [api-docs](https://www.npmjs.com/package/@backstage/plugin-api-docs) plugin which is responsible for indexing the API specification content surfaced via the API tab of a catalog entity.

The plugin provides an `ApiDocumentCollatorFactory` which is an implementation of the `DocumentCollatorFactory`. Once integrated into the search plugin, users will be able to search API specification content and navigate to the corresponding API definition.

Screen Shot 2022-10-20 at 4 49 18 PM

Screen Shot 2022-10-20 at 4 23 30 PM

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by iamEAP — COMMENTED** [review]
Thanks for the contrib, @jolavb!  Before going too deep in the details, I took a look at the API surface area; some minor things here and there, but the main thing I'd love your thoughts on is the `SpecHandler`.

File: plugins/search-backend-module-api-docs/README.md
Comment (jolavb) [reply plugins/search-backend-module-api-docs/README.md:1]: 👍 Curious where you think it'd make sense to include these docs- following the pattern I see in the existing docs, maybe I add it to the marketplace (which can link to the readme) and then a link to the marketplace plugin [here](https://backstage.io/docs/features/search/search-overview#plugins-integrated-with-backstage-search)

Comment (awanlin) [reply plugins/search-backend-module-api-docs/README.md:1]: Would it not make sense to add it here? https://backstage.io/docs/features/search/how-to-guides#how-to-index-techdocs-documents Feels a lot like what's being done for TechDocs. 

Sorry, been lurking on this PR as I know this is something we will for sure take advantage of once it gets merged and released 🚀

Comment (iamEAP) [reply plugins/search-backend-module-api-docs/README.md:1]: That'd be a good spot for now! We could figure out a better place if/when we expand with more detail (eg custom spec parsers, etc)

File: plugins/search-backend-module-api-docs/package.json
Comment (camilaibs) [inline plugins/search-backend-module-api-docs/package.json:33]: It seems that this dependency is not used

**Review by camilaibs — COMMENTED** [review]
Thanks for contributing with a new plugin, so nice of you ⭐ !
I commented on with some ideas, let me know what you think:

File: plugins/search-backend-module-api-docs/src/search/spec-parsers/OpenAPISpecParser.ts
Comment (iamEAP) [reply plugins/search-backend-module-api-docs/src/search/spec-parsers/OpenAPISpecParser.ts:23]: Yeah, it's easy to imagine this evolving to have many implementations and uses outside of just the search collator.  But perhaps it's best to start small and keep it in here?  Multiple specs could evolve in-place, and then if/when there's a use for them outside of this package, they could be broken out.

Maybe it'd be worth marking this internal instead of public, even?

File: plugins/search-backend-module-api-docs/src/search/collators/APIDocumentCollatorFactory.ts
Comment (iamEAP) [reply plugins/search-backend-module-api-docs/src/search/collators/APIDocumentCollatorFactory.ts:93]: (This should be handled in all of the search engines now: https://github.com/backstage/backstage/pull/14938)

File: plugins/search-backend-module-api-docs/api-report.md
Comment (jolavb) [reply plugins/search-backend-module-api-docs/api-report.md:83]: Yeah I'd say that's a safe bet.

File: plugins/search-backend-module-api-docs/README.md
Comment (jolavb) [reply plugins/search-backend-module-api-docs/README.md:1]: 👍 Added some docs under [https://backstage.io/docs/features/search/how-to-guides](https://backstage.io/docs/features/search/how-to-guides#how-to-index-techdocs-documents)

File: plugins/search-backend-module-api-docs/api-report.md
Comment (jolavb) [reply plugins/search-backend-module-api-docs/api-report.md:78]: I'm not familiar with UrlReader or their friends 😃  but I'm on board with abstracting away the rest of `SpecHandler` - no need to make folks re-implement addSpecParser and getSpecParser.

Comment (jolavb) [reply plugins/search-backend-module-api-docs/api-report.md:62]: Added `SpecParser[]` on the collator factory options 👍

File: plugins/search-backend-module-api-docs/api-report.md
Comment (jolavb) [reply plugins/search-backend-module-api-docs/api-report.md:24]: Agreed!

File: plugins/search-backend-module-api-docs/api-report.md
Comment (jolavb) [reply plugins/search-backend-module-api-docs/api-report.md]: Sure thing 👍

**Comment by jolavb** [comment]
> Thanks for contributing with a new plugin, so nice of you ⭐ ! I commented on with some ideas, let me know what you think:

Thanks for your review- getting to these comments one by one as time allows. Typescript is not my first language and I have limited backstage dev experience so I'm all ears on suggestions & improvements.

File: plugins/search-backend-module-api-docs/api-report.md
Comment (jolavb) [reply plugins/search-backend-module-api-docs/api-report.md:20]: Right now we're returning the following by default:

```
title: entity.metadata.name,
location: `/catalog/default/api/${entity.metadata.name}/definition/`,
text: specParser.getSpecText(entity.spec?.definition as string),
kind: entity.kind,
lifecycle: (entity.spec?.lifecycle as string) || '',
```

I could maybe see folks wanting to see extra fields, but I'm maybe that should be split out into it's own PR? Also, curious if we actually need. `APIDefinitionDocument`, since it pretty much overlaps with `CatalogEntityDocument`. Maybe if folks would want to see something specific to an APIDefinition like 'provided by` ?

File: plugins/search-backend-module-api-docs/api-report.md
Comment (jolavb) [reply plugins/search-backend-module-api-docs/api-report.md]: Yeah, [iamEAP](https://github.com/iamEAP) called that out in their review and wound up changing this to an array of `SpecParsers`. I almost considered not making this configurable at all though- I would hope that as folks see a need for a new type of parser for things that are supported by api-definition plugin (graphql, gprc for instance) , it gets contributed back for all of us to take advantage of and added to the default list of parsers.

File: plugins/search-backend-module-api-docs/api-report.md
Comment (jolavb) [reply plugins/search-backend-module-api-docs/api-report.md]: Reasoning here was just that I wanted to clearly separate the logic between parsing specs of different versions. I think accepting a version param and then handling individual fields based on version could get pretty hard to read quickly- especially as new versions of the spec are released. This limits the amount of branching logic required of a single function and allows for an easier read.

File: plugins/search-backend-module-api-docs/src/search/collators/APIDocumentCollatorFactory.ts
Comment (jolavb) [reply plugins/search-backend-module-api-docs/src/search/collators/APIDocumentCollatorFactory.ts]: Good call!

File: plugins/search-backend-module-api-docs/package.json
Comment (jolavb) [reply plugins/search-backend-module-api-docs/package.json:39]: Thanks- forgot to clean up package.json

**Comment by ollid2405** [comment]
My company is exited to see this plugin ship.
@jolavb  - Do you still have plans to push this along?

**Comment by jolavb** [comment]
> My company is exited to see this plugin ship. @jolavb - Do you still have plans to push this along?

Hi @ollid2405 Yeah would love to get this out the door. @camilaibs and @iamEAP any additional input/advice on your reviews? Not sure what state everyone's in re: all these layoffs :(  I for one...suddenly have plenty of time on my hands.

**Comment by Rugvip** [comment]
@jolavb could you make sure all commits are signed with DCO so that we can have a look 🙏

**Comment by jolavb** [comment]
> @jolavb could you make sure all commits are signed with DCO so that we can have a look 🙏

Yes apologies, I hadn't realized I was missing DCO in a few commits. Currently working on getting push access to `wayfair-contribs` restored post-layoff.

**Comment by Rugvip** [comment]
@jolavb ah shit. no rush with this. It might be smoother to close this PR and re-open when ready from a different fork

File: plugins/search-backend-module-api-docs/api-report.md
Comment (camilaibs) [reply plugins/search-backend-module-api-docs/api-report.md:20]: At the moment, nothing comes to mind. It'd be great if the @backstage/maintainers would give their opinion. I'd also suggest sharing on discord so we can get ideas from others 🙂 .

File: plugins/search-backend-module-api-docs/api-report.md
Comment (camilaibs) [reply plugins/search-backend-module-api-docs/api-report.md]: Sounds good!

File: plugins/search-backend-module-api-docs/api-report.md
Comment (camilaibs) [reply plugins/search-backend-module-api-docs/api-report.md]: My mind was more on the public interface, requiring only one method to be invoked, passing the version. Internally, this method can call the separate methods based on the version, making them private...

**Review by camilaibs — APPROVED** [review]
Thanks for contributing, looks great to me 🎉

**Comment by ollid2405** [comment]
Hi @jolavb, are you working on another pr at this feature? 
I can try to help you to get this shipped, but I am not that experienced in Backstage development 😕

**Pull request closed without merging**
